include/deal.II/integrators/l2.h

```diff
@@ -63,17 +63,21 @@ namespace LocalIntegrators
           const double dx = fe.JxW(k) * factor;
 
           for (unsigned int i=0; i<n_dofs; ++i)
-            for (unsigned int j=i; j<n_dofs; ++j)
-              for (unsigned int d=0; d<n_components; ++d)
-                M(i,j) += dx
-                          * fe.shape_value_component(j,k,d)
+            for (unsigned int d=0; d<n_components; ++d)
+              {
+                M(i,i) += dx
+                          * fe.shape_value_component(i,k,d)
                           * fe.shape_value_component(i,k,d);
+                for (unsigned int j=i+1; j<n_dofs; ++j)
+                  {
+                    const double Mij =  dx
+                                        * fe.shape_value_component(j,k,d)
+                                        * fe.shape_value_component(i,k,d);
+                    M(i,j) += Mij;
+                    M(j,i) += Mij;
+                  }
+              }
```

Comment:
Why not accumulate the sum over n_components (for both Mii and Mij)? This way, you should benefit a bit more if that is larger than one.